src/net/client/cache.rs

```diff
@@ -0,0 +1,1177 @@
+        if value.valid_for.is_zero() {
+            // Do not insert cache value that are valid for zero duration.
+            return;
+        }
+        let value = match prepare_for_insert(value.clone(), &self.config) {
+            Ok(value) => value,
+            Err(e) => {
+                // Create a new value based on this error
+                Arc::new(
+                    Value::<C>::new_from_value_and_response(
+                        value,
+                        Err(e),
+                        &self.config,
+                    )
+                    .expect("value from error does not fail"),
+                )
+            }
+        };
+        self.cache.insert(key, value).await
+    }
+}
+
+impl<CR, Upstream, C> Debug for Request<CR, Upstream, C>
+where
+    CR: Send + Sync,
+    Upstream: Send + Sync,
+    C: Clock + Send + Sync,
+{
+    fn fmt(&self, f: &mut Formatter<'_>) -> Result<(), core::fmt::Error> {
+        f.debug_struct("Request")
+            .field("fut", &format_args!("_"))
+            .finish()
+    }
+}
+
+impl<CR, Upstream, C> GetResponse for Request<CR, Upstream, C>
+where
+    CR: Clone + ComposeRequest + Debug + Sync,
+    Upstream: SendRequest<CR> + Send + Sync + 'static,
+    C: Clock + Debug + Send + Sync + 'static,
+{
+    fn get_response(
+        &mut self,
+    ) -> Pin<
+        Box<
+            dyn Future<Output = Result<Message<Bytes>, Error>>
+                + Send
+                + Sync
+                + '_,
+        >,
+    > {
+        Box::pin(self.get_response_impl())
+    }
+}
+
+//------------ RequestState ---------------------------------------------------
+/// States of the state machine in get_response_impl
+enum RequestState {
+    /// Initial state, perform a cache lookup.
+    Init,
+
+    /// Wait for a response and insert the response in the cache.
+    GetResponse(Key, Box<dyn GetResponse + Send + Sync>),
+
+    /// Wait for a response but do not insert the response in the cache.
+    GetResponseNoCache(Box<dyn GetResponse + Send + Sync>),
+}
+
+//------------ Key ------------------------------------------------------------
+
+#[derive(Clone, Debug, Eq, Hash, PartialEq)]
+/// The key for cache entries.
+struct Key {
+    /// DNS name in the request.
+    qname: Dname<Vec<u8>>,
+
+    /// The request class. Always IN at the moment.
+    qclass: Class,
+
+    /// The requested type.
+    qtype: Rtype,
+
+    /// Value of the AD flag.
+    ad: bool,
+
+    /// Value of the CD flag.
+    cd: bool,
+
+    /// Value of the DO flag.
+    dnssec_ok: bool,
+
+    /// Value of the RD flag.
+    rd: bool,
+}
+
+impl Key {
+    /// Create a new key object.
+    fn new<TDN>(
+        qname: TDN,
+        qclass: Class,
+        qtype: Rtype,
+        ad: bool,
+        cd: bool,
+        dnssec_ok: bool,
+        rd: bool,
+    ) -> Key
+    where
+        TDN: ToDname,
+    {
+        let mut qname = qname.to_dname().expect("to_dname should not fail");
```

Comment:
We should probably switch `ToDname` to have a pair of `try_to_dname` and `to_dname` methods to avoid these kinds of “unwraps.” Most other places already have this.

Reply:
Doesn't seem to be an action on this PR.

Reply:
I created #281